### ISSUE TITLE
ma12-45

### ISSUE THREAD

**Issue by dblodgett-usgs** [USGS-R/EflowStats#125]
ma12-45 are all off by significant, but not massive, amounts. My guess is that it's an issue of water vs calendar year.

These are all dependent on monthly flow statistics.

**Comment by dblodgett-usgs** [comment]
Looking at 12 to 24, I thought maybe the months got out of order. But pulling out one site and looking at the old and new, none of the values match eachother. 
```
> new_results2["01406000", 1:11]
           ma1 ma2    ma3    ma4   ma5  ma6   ma7   ma8   ma9  ma10  ma11
01406000 13.97 9.4 143.78 82.953 1.486 10.4 4.359 3.125 1.045 0.657 0.509
> old_results2["01406000", 1:11]
           ma1 ma2     ma3    ma4   ma5  ma6   ma7   ma8   ma9  ma10  ma11
01406000 13.97 9.4 130.956 38.064 1.486 10.4 4.359 3.125 1.045 0.657 0.509
```
Given that many of the ma stats are correct. I have to chalk this up to differences in how these are calculated. But what would be different? I don't see how this could be from rounding.
```
> new_results2["01406000", 12:24]
           ma12   ma13  ma14  ma15   ma16  ma17 ma18  ma19   ma20  ma21   ma22   ma23   ma24
01406000 18.075 18.075 22.95 20.86 12.753 10.56 7.68 5.647 14.105 9.195 14.784 13.466 92.633
> old_results2["01406000", 12:24]
           ma12   ma13   ma14  ma15   ma16  ma17  ma18  ma19   ma20  ma21   ma22   ma23   ma24
01406000 18.256 16.041 23.584 21.28 11.805 7.945 4.582 5.644 10.432 8.574 13.708 12.348 90.867
```

**Comment by dblodgett-usgs** [comment]
Added a test to verify that months are getting pulled out right. They are. Not sure what else to do here.

**Comment by dblodgett-usgs** [comment]
The plot thickens. Looking at ma36 to 45, many are spot on in the new implementation and they all appear to be within rounding error?
> new_results2["01406000", 36:45]
          ma36  ma37  ma38   ma39 ma40  ma41  ma42  ma43  ma44   ma45
01406000 2.844 1.075 1.803 60.876 0.15 1.731 0.344 0.225 0.344 -0.007
> old_results2["01406000", 36:45]
          ma36  ma37  ma38   ma39 ma40  ma41  ma42  ma43  ma44   ma45
01406000 2.844 1.075 1.803 60.876 0.15 1.729 0.341 0.228 0.341 -0.009

So another theory worth investigating is if the old implementation rounded all the monthly means prior to taking the mean of all of them.

**Comment by dblodgett-usgs** [comment]
Nope. I did this:
```
        x$my <- paste(x$month_val, x$year_val, sep="-")
        
        flowSum_my <- dplyr::summarize(dplyr::group_by(x,my),
                                       monthlyMeanFlow = round(mean.default(discharge),digits=0),
                                       month_val = mean(month_val))
        
        ma12.23 <- dplyr::summarize(dplyr::group_by(flowSum_my,month_val),
                                    meanFlow = mean.default(monthlyMeanFlow))$meanFlow
```
instead of this:
```
                flowSum_month <- dplyr::summarize(dplyr::group_by(x,month_val),
                                                  meanFlow = mean.default(discharge))
                ma12.23 <- flowSum_month$meanFlow
```

and tried digits=0, and 1. The changes are smaller than the differences.

**Comment by dblodgett-usgs** [comment]
42-45 are darn close. Such a mystery why they would be off by these amounts though.
```
> new_results2["01406000", 42:45]
          ma42  ma43  ma44   ma45
01406000 0.344 0.225 0.344 -0.007
> old_results2["01406000", 42:45]
          ma42  ma43  ma44   ma45
01406000 0.341 0.228 0.341 -0.009
```

**Comment by dblodgett-usgs** [comment]
#124 includes all the code changes that should "fix" this issue. A number of things have not actually been fixed, but I think all our calculations DO match the definitions.

**Comment by dblodgett-usgs** [comment]
This is bizarre. There must be a bug in the old (2006) code with calculating mean monthly summary values. I found that the monthly period of record summary statistics contain a value from a particular monthly mean from the monthly mean time series. 

In the following code, temp.rda has `new`, `old`, and `flow_data_clean` for one site time series.  
`my` is calculated to contain mean monthly values for all months in the dataset.  
The big ugly loop searches for matches between the mean monthly values and the monthly mean summary values in the output. Results of the code are pasted below the code block.
```
library(dplyr)
load("data/temp.rda")

flow_data_clean$month_val <- lubridate::month(flow_data_clean$date)

flow_data_clean$my <- paste(flow_data_clean$month_val, flow_data_clean$year_val, sep="-")

my <- flow_data_clean %>% group_by(my) %>%
        summarise(monthmean = mean(discharge))

# ma12 - 23
start <- c(1, 12, 23, 34, 45, 56, 67, 78, 89, 100, 111, 122)
end <- start[2:12] - 1
end <- c(end, 132)
expect <- old[,c(12, 21, 22, 23, 13, 14, 15, 16, 17, 18, 19, 20)]
for(s in 1:12) { 
        for(i in start[s]:end[s]) {
                for(j in start[s]:end[s]) {
                        if(abs(mean(my$monthmean[i:j]) - expect[s][[1]])<0.002) {
                                print(paste(i, 
                                            j, 
                                            round(mean(my$monthmean[i:j]), digits = 3), 
                                            expect[s]))
                        }
                        
                }
        }
}

[1] "6 6 11.513 11.513"
[1] "19 19 6.177 6.177"
[1] "30 30 7.423 7.423"
[1] "42 42 7.835 7.835"
[1] "51 51 10.868 10.868"
[1] "65 65 17.516 17.516"
[1] "76 76 18.877 18.877"
[1] "82 82 14.677 14.677"
[1] "89 89 9.34 9.34"
[1] "105 105 7.758 7.758"
[1] "117 117 7.365 7.365"
[1] "127 127 5.397 5.397"
```

This shows that a single value matches the value in the processed result from the old (2006) code.

For some reason, a particular mean from one month is in the output rather than the mean of all the months. 

Basically, this means that we can't say much about the correspondence between EflowStats and this output.

Going to close this issue and move on to comparing some that have larger discrepancies.
